pkg/controller/utils/ownership_test.go

```diff
@@ -0,0 +1,185 @@
+				AWS: &hivev1.AWSDNSZoneStatus{
+					ZoneID: aws.String("1234"),
+				},
+			},
+		}
+
+		return zone
+	}
+
+	validDNSZoneWithMissingOwnership = func(owner hivev1.MetaRuntimeObject, scheme *runtime.Scheme) *hivev1.DNSZone {
+		zone := validDNSZoneWithLabelOwner(owner, scheme)
+		return zone
+	}
+
+	validDNSZoneWithControllerOwnership = func(labelOwner, controllerOwner hivev1.MetaRuntimeObject, scheme *runtime.Scheme) *hivev1.DNSZone {
+		zone := validDNSZoneWithLabelOwner(labelOwner, scheme)
+		controllerutil.SetControllerReference(controllerOwner, zone, scheme)
+		return zone
+	}
+)
+
+func TestReconcile(t *testing.T) {
+	testscheme := scheme.Scheme
+	apis.AddToScheme(testscheme)
+
+	tests := []struct {
+		name                string
+		owner               hivev1.MetaRuntimeObject
+		ownershipUniqueKeys []*OwnershipUniqueKey
+		existingObjects     []runtime.Object
+		expectedObjects     []runtime.Object
+	}{
+		{
+			name:                "no objects in kube (do nothing)",
+			owner:               validClusterDeployment(),
+			ownershipUniqueKeys: validOwnershipUniqueKeys(validClusterDeployment()),
+			existingObjects:     emptyRuntimeObjectSlice(),
+			expectedObjects:     emptyRuntimeObjectSlice(),
+		},
+		{
+			name:                "ownership set correctly (do nothing)",
+			owner:               validClusterDeployment(),
+			ownershipUniqueKeys: validOwnershipUniqueKeys(validClusterDeployment()),
+			existingObjects: []runtime.Object{
+				validDNSZoneWithControllerOwnership(validClusterDeployment(), validClusterDeployment(), testscheme),
+			},
+			expectedObjects: []runtime.Object{
+				validDNSZoneWithControllerOwnership(validClusterDeployment(), validClusterDeployment(), testscheme),
+			},
+		},
+		{
+			name:                "ownership missing (add ownership back)",
+			owner:               validClusterDeployment(),
+			ownershipUniqueKeys: validOwnershipUniqueKeys(validClusterDeployment()),
+			existingObjects: []runtime.Object{
+				validDNSZoneWithMissingOwnership(validClusterDeployment(), testscheme),
+			},
+			expectedObjects: []runtime.Object{
+				validDNSZoneWithControllerOwnership(validClusterDeployment(), validClusterDeployment(), testscheme),
+			},
+		},
+		{
+			name:                "ownership incorrect (fix it)",
+			owner:               validClusterDeployment(),
+			ownershipUniqueKeys: validOwnershipUniqueKeys(validClusterDeployment()),
+			existingObjects: []runtime.Object{
+				validDNSZoneWithControllerOwnership(validClusterDeployment(), validOtherClusterDeployment(), testscheme),
+			},
+			expectedObjects: []runtime.Object{
+				validDNSZoneWithControllerOwnership(validClusterDeployment(), validClusterDeployment(), testscheme),
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Arrange
+			fakeKubeClient := fake.NewFakeClient(test.existingObjects...)
+			logger := log.WithField("fake", "fake")
+			ownershipKey := test.ownershipUniqueKeys[0]
```

Comment:
It may be better to have the test cases use a `ownershipUniqueKey *OwnershipUniqueKey` instead of a slice. Otherwise, there is an implicit requirement on the test cases that there be exactly one element in the slice.